File: Sliding_Window/code/Solution_1/sol.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

int lengthOfLongestSubstring(string s) {
    unordered_map<char, int> lastSeen;
    int left = 0, maxLen = 0;

    for (int right = 0; right < s.size(); ++right) {
        char c = s[right];
        if (lastSeen.count(c)) {
            left = max(left, lastSeen[c] + 1);
        }
        lastSeen[c] = right;
        maxLen = max(maxLen, right - left + 1);
    }

    return maxLen;
}

int main() {
    string s1 = "abcabcbb";
    string s2 = "bbbbb";
    string s3 = "pwwkew";

    cout << lengthOfLongestSubstring(s1) << endl; // 3
    cout << lengthOfLongestSubstring(s2) << endl; // 1
    cout << lengthOfLongestSubstring(s3) << endl; // 3

    return 0;
}

File: Sliding_Window/code/Solution_1/sol.py
def length_of_longest_substring(s: str) -> int:
    last_seen = {}
    left, max_len = 0, 0

    for right, char in enumerate(s):
        if char in last_seen:
            left = max(left, last_seen[char] + 1)
        last_seen[char] = right
        max_len = max(max_len, right - left + 1)

    return max_len

if __name__ == "__main__":
    s1 = "abcabcbb"
    s2 = "bbbbb"
    s3 = "pwwkew"

    print(length_of_longest_substring(s1))  # 3
    print(length_of_longest_substring(s2))  # 1
    print(length_of_longest_substring(s3))  # 3
